include/apq/query.h

```diff
@@ -1,12 +1,184 @@
-#pragma once
+#include <apq/binary_serialization.h>
+#include <apq/type_traits.h>
 
-#include <boost/hana.hpp>
+#include <boost/hana/for_each.hpp>
+#include <boost/hana/tuple.hpp>
+
+#include <libpq-fe.h>
+
+#include <array>
+#include <iterator>
+#include <memory>
+#include <optional>
+#include <vector>
 
 namespace libapq {
 
-template <const char* text, class ... Args>
-auto make_query(Args&& ... args) {
-    return boost::hana::make_tuple(std::forward<Args&&>(args) ...);
+template <class T>
+using not_nullopt_t = std::enable_if_t<!std::is_same_v<T, std::nullopt_t>, T>;
+
+template <class T>
+using is_nullable_t = std::enable_if_t<is_nullable<std::decay_t<T>>::value, void>;
+
+template <class T>
+using is_not_nullable_t = std::enable_if_t<!is_nullable<std::decay_t<T>>::value, void>;
+
```

Comment:
A minute of advertising. :)

With Require from connection branch this can be rewritten like:

    template <typename T>
    constexpr auto Nullable = is_nullable<std::decay_t<T>>::value;

...

    template <class T>
    static constexpr Require<!Nullable<T>> write_meta(const oid_map_type& oid_map, ...)

Reply:
Done